test/test_cuda_sanitizer.py

```diff
@@ -0,0 +1,433 @@
+        b = torch.empty(8, device="cuda")
+
+        argument_handler = csan.ArgumentHandler()
+        argument_handler.parse_inputs(mul_out_func._schema, (a, 3), {"out": b})
+        torch.mul(a, 3, out=b)
+        argument_handler.parse_outputs(b)
+
+        self.assertEqual({a.data_ptr()}, argument_handler.dataptrs_read)
+        self.assertEqual({b.data_ptr()}, argument_handler.dataptrs_written)
+
+    def test_nonzero(self):
+        nonzero_func = torch.ops.aten.nonzero.default
+        a = torch.ones(5, 3, 2, device="cuda")
+
+        argument_handler = csan.ArgumentHandler()
+        argument_handler.parse_inputs(nonzero_func._schema, (a,), {"as_tuple": True})
+        out = torch.nonzero(a, as_tuple=True)
+        argument_handler.parse_outputs(out)
+
+        outputs = {out[0].data_ptr(), out[1].data_ptr(), out[2].data_ptr()}
+        self.assertEqual({a.data_ptr()}, argument_handler.dataptrs_read)
+        self.assertEqual(outputs, argument_handler.dataptrs_written)
+
+    def test_tensor_names(self):
+        addr_func = torch.ops.aten.addr.default
+        vec = torch.arange(1, 4, device="cuda")
+        M = torch.zeros(3, 3, device="cuda")
+
+        argument_handler = csan.ArgumentHandler()
+        argument_handler.parse_inputs(addr_func._schema, (M, vec, vec), {})
+        out = torch.addr(M, vec, vec)
+        argument_handler.parse_outputs(out)
+
+        self.assertEqual(
+            argument_handler.tensor_names,
+            {
+                M.data_ptr(): ["self"],
+                vec.data_ptr(): ["vec1", "vec2"],
+                out.data_ptr(): ["output"],
+            },
+        )
+
+
+def tensor_id(i: int) -> DataPtr:
+    return i
+
+
+def stream_id(i: int) -> StreamId:
+    return 1000 + i
+
+
+def event_id(i: int) -> EventId:
+    return 2000 + i
+
+
+class TestEventHandler(TestCase):
+    def setUp(self):
+        self.handler = csan.EventHandler()
+
+    def kernel_launch(
+        self,
+        stream: StreamId,
+        read_only: List[DataPtr] = None,
+        read_write: List[DataPtr] = None,
+    ) -> List[csan.SynchronizationError]:
+        if read_only is None:
+            read_only = []
+        if read_write is None:
+            read_write = []
+        return self.handler._handle_kernel_launch(
+            stream,
+            read_only,
+            read_write,
+            "",
+            {k: [""] for k in read_only + read_write},
+        )
+
+    def assert_good_kernel_launch(
+        self,
+        stream: StreamId,
+        read_only: List[DataPtr] = None,
+        read_write: List[DataPtr] = None,
+    ) -> None:
+        self.assertEqual(self.kernel_launch(stream, read_only, read_write), [])
+
+    def assert_bad_kernel_launch(
+        self,
+        stream: StreamId,
+        read_only: List[DataPtr] = None,
+        read_write: List[DataPtr] = None,
+    ) -> None:
+        self.assertNotEqual(self.kernel_launch(stream, read_only, read_write), [])
```

Comment:
what result do you expect here? I guess it would depend on what you are testing, but you can make expected result an argument to assert to do a more exact check

Reply:
I've added an argument to test the number of errors reported.